pkg/agent/globalstate.go

```diff
@@ -172,9 +146,170 @@ func (s *agentState) handleVMEventAdded(
+	// Three steps:
+	//  1. Check if the Runner needs to restart. If no, we're done.
+	//  2. Wait for a random amount of time (between RunnerRestartMinWaitSeconds and RunnerRestartMaxWaitSeconds)
+	//  3. Restart the Runner (if it still should be restarted)
+
+	pod, ok := func() (*podState, bool) {
+		// note: intentionally release s.lock as soon as we're done with it, so that there's no
+		// possibility of holding onto it longer than we need.
+		s.lock.Lock()
+		defer s.lock.Unlock()
+		pod, ok := s.pods[podName]
+		return pod, ok
+	}()
+
+	if !ok {
+		return
+	}
+
+	pod.status.mu.Lock()
+	defer pod.status.mu.Unlock()
+
+	if pod.status.endState != nil {
+		klog.Errorf(
+			"TriggerRestartIfNecessary called with nil endState for pod %v (should only be called after the pod is finished, when endState != nil)",
+			podName,
+		)
+		s.metrics.runnerFatalErrors.Inc()
+		return
+	}
+
+	endTime := pod.status.endState.Time
+
+	if endTime.IsZero() {
+		// If we don't check this, we run the risk of spinning on failures.
+		klog.Errorf("TriggerRestartIfNecessary called with zero'd Time for pod %v", podName)
+		s.metrics.runnerFatalErrors.Inc()
+		// Continue on, but with the time overridden, so we guarantee our minimum wait.
+		endTime = time.Now()
+	}
+
+	// keep this for later.
+	exitKind := pod.status.endState.ExitKind
+
+	switch exitKind {
+	case podStatusExitCanceled:
+		return // successful exit, no need to restart.
+	case podStatusExitPanicked, podStatusExitErrored:
+		// Should restart; continue.
+	default:
+		klog.Errorf("TriggerRestartIfNecessary called with unexpected ExitKind %q for pod %v", podName)
+		s.metrics.runnerFatalErrors.Inc()
+		return
+	}
+
+	// Begin steps (2) and (3) -- wait, then restart.
+	var waitDuration time.Duration
+	totalRuntime := endTime.Sub(pod.status.startTime)
+
+	// If the runner was running for a while, restart immediately.
+	//
+	// NOTE: this will have incorrect behavior when the system clock is behaving weirdly, but that's
+	// mostly ok. It's ok to e.g. restart an extra time at the switchover to daylight saving time.
+	if totalRuntime > time.Second*time.Duration(RunnerRestartMaxWaitSeconds) {
+		waitDuration = 0
+	} else /* Otherwise, randomly pick within RunnerRestartMinWait..RunnerRestartMaxWait */ {
+		r := util.NewTimeRange(time.Second, RunnerRestartMinWaitSeconds, RunnerRestartMaxWaitSeconds)
+		waitDuration = r.Random()
+		klog.Info("Waiting for %s before restarting Runner %v", waitDuration, podName)
+	}
+
+	// Run the waiting (if necessary) and restarting in another goroutine, so we're not blocking the
+	// caller of this function.
+	go func() {
+		if waitDuration != 0 {
+			select {
+			case <-time.After(waitDuration):
+			case <-runnerCtx.Done():
+				klog.Infof("Canceling restart of Runner %v after %s: %s", podName, time.Since(endTime), runnerCtx.Err())
+				return
+			}
+		}
+
+		s.lock.Lock()
+		defer s.lock.Unlock()
```

Comment:
why don't we do the narrow scope lock situation here? 

also we still have a pointer to pod (which should be the same object) in scope because of like 165.

Reply:
Added a comment:

> Need to update pod itself; can't release s.lock. Also, pod *theoretically* may been deleted + restarted since we started, so it's incorrect to hold on to the original podStatus.

lmk if that works